train.py

```diff
@@ -22,88 +22,182 @@ def _avoid_overwriting(experiment_id):
+                'lr': params["optimization"]["transformer_lr"]
+            },
+            {
+                'params': self.hgn.hnn.parameters(),
+                'lr': params["optimization"]["hnn_lr"]
+            },
+            {
+                'params': self.hgn.decoder.parameters(),
+                'lr': params["optimization"]["decoder_lr"]
+            },
+        ]
+        self.optimizer = torch.optim.Adam(optim_params)
+
+    def training_step(self, rollouts):
+        """Perform a training step with the given rollouts batch.
+
+        TODO: Move the whole fit() code inside forward if adding hooks to the training, as direct
+            calls to self.forward() do not trigger them.
+
+        Args:
+            rollouts (torch.Tensor): Tensor of shape (batch_size, seq_len, channels, height, width)
+                corresponding to a batch of sampled rollouts.
+            variational (bool): Whether to sample from the encoder and compute the KL loss,
+                or train in fully deterministic mode.
+
+        Returns:
+            A dictionary of losses and the model's prediction of the rollout. The reconstruction loss and
+            KL divergence are floats and prediction is the HGNResult object with data of the forward pass.
+        """
+
+        hgn_output = self.hgn.forward(rollout_batch=rollouts)
+        target = hgn_output.input
+        prediction = hgn_output.reconstructed_rollout
+        
+        if self.params["networks"]["variational"]:    
+            C, rec_loss = geco_constraint(target, prediction, self.params["geco"]["tol"])
+            C_curr = C.item()
+            # Compute KL divergence
+            mu = hgn_output.z_mean
+            logvar = hgn_output.z_logvar
+            kld = kld_loss(mu, logvar)
+    
+            # Compute moving average of constraint C
+            if self.C_ma is None:
+                self.C_ma = C
+            else:
+                # not exactly sure if i should detach here on in the expression below 
+                self.C_ma = self.params["geco"]["alpha"] * self.C_ma.detach() + \
+                            (1 - self.params["geco"]["alpha"]) * C
+
+            C = C + (self.C_ma - C)
+            
+            # Compute losses
+            train_loss = kld + self.langrange_multiplier * C
+
+            losses = {'loss/train': train_loss,
+                      'loss/kld': kld,
+                      'loss/C_cur': C_curr,
+                      'loss/C_ma': self.C_ma,
+                      'loss/rec': rec_loss,
+                      'other/langrange_mult': self.langrange_multiplier
+                     }
+
+            # clamping the langrange multiplier to avoid inf values
+            self.langrange_multiplier = torch.clamp(self.langrange_multiplier * 
+                                                    torch.exp(self.params["geco"]["langrange_multiplier_param"]
+                                                    * self.C_ma.detach()), 1e-10, 1e10)
```

Comment:
Same as before, I think you want C_ma detached from the beginning. You can use C.detach() to update it

Reply:
See prev answer.